File: oauth2/typescript/src/flows/device.ts
/**
 * Device Authorization Flow
 *
 * RFC 8628 - OAuth 2.0 Device Authorization Grant.
 */

import {
  DeviceCodeParams,
  DeviceAuthorizationResponse,
  DeviceTokenResult,
  TokenResponse,
  OAuth2Config,
} from "../types";
import { OAuth2Error, AuthorizationError, ProviderError, ConfigurationError } from "../error";
import { createErrorFromResponse, parseErrorResponse } from "../error/mapping";
import { HttpTransport } from "../core/transport";

/**
 * Device Authorization Flow interface.
 */
export interface DeviceAuthorizationFlow {
  /**
   * Request device and user codes.
   */
  requestDeviceCode(params?: DeviceCodeParams): Promise<DeviceAuthorizationResponse>;

  /**
   * Poll for token using device code.
   */
  pollToken(deviceCode: string): Promise<DeviceTokenResult>;

  /**
   * Await authorization by polling until success/failure/timeout.
   */
  awaitAuthorization(
    response: DeviceAuthorizationResponse,
    options?: { maxWaitSeconds?: number }
  ): Promise<TokenResponse>;
}

/**
 * Device Authorization Flow implementation.
 */
export class DeviceAuthorizationFlowImpl implements DeviceAuthorizationFlow {
  private config: OAuth2Config;
  private transport: HttpTransport;

  constructor(config: OAuth2Config, transport: HttpTransport) {
    this.config = config;
    this.transport = transport;

    // Validate device authorization endpoint exists
    if (!config.provider.deviceAuthorizationEndpoint) {
      throw new ConfigurationError(
        "Device authorization endpoint not configured",
        "MissingRequired"
      );
    }
  }

  async requestDeviceCode(params?: DeviceCodeParams): Promise<DeviceAuthorizationResponse> {
    const body = new URLSearchParams();
    body.set("client_id", this.config.credentials.clientId);

    // Scopes
    const scopes = params?.scopes ?? this.config.defaultScopes;
    if (scopes && scopes.length > 0) {
      body.set("scope", scopes.join(" "));
    }

    const headers: Record<string, string> = {
      "content-type": "application/x-www-form-urlencoded",
      accept: "application/json",
    };

    const response = await this.transport.send({
      method: "POST",
      url: this.config.provider.deviceAuthorizationEndpoint!,
      headers,
      body: body.toString(),
      timeout: this.config.timeout,
    });

    if (response.status !== 200) {
      throw createErrorFromResponse(response.status, response.body);
    }

    return this.parseDeviceAuthorizationResponse(response.body);
  }

  async pollToken(deviceCode: string): Promise<DeviceTokenResult> {
    const body = new URLSearchParams();
    body.set("grant_type", "urn:ietf:params:oauth:grant-type:device_code");
    body.set("device_code", deviceCode);
    body.set("client_id", this.config.credentials.clientId);

    const headers: Record<string, string> = {
      "content-type": "application/x-www-form-urlencoded",
      accept: "application/json",
    };

    // Add client secret if confidential client
    if (this.config.credentials.clientSecret) {
      if (this.config.credentials.authMethod === "client_secret_basic") {
        const credentials = Buffer.from(
          `${this.config.credentials.clientId}:${this.config.credentials.clientSecret}`
        ).toString("base64");
        headers["authorization"] = `Basic ${credentials}`;
      } else {
        body.set("client_secret", this.config.credentials.clientSecret);
      }
    }

    const response = await this.transport.send({
      method: "POST",
      url: this.config.provider.tokenEndpoint,
      headers,
      body: body.toString(),
      timeout: this.config.timeout,
    });

    // Success
    if (response.status === 200) {
      const tokens = this.parseTokenResponse(response.body);
      return { status: "success", tokens };
    }

    // Check for error response
    const errorResponse = parseErrorResponse(response.body);
    if (errorResponse) {
      switch (errorResponse.error) {
        case "authorization_pending":
          return { status: "pending" };
        case "slow_down":
          return { status: "slow_down", newInterval: 5 };
        case "expired_token":
          return { status: "expired" };
        case "access_denied":
          return { status: "access_denied" };
        default:
          throw createErrorFromResponse(response.status, response.body);
      }
    }

    throw createErrorFromResponse(response.status, response.body);
  }

  async awaitAuthorization(
    response: DeviceAuthorizationResponse,
    options?: { maxWaitSeconds?: number }
  ): Promise<TokenResponse> {
    const maxWait = (options?.maxWaitSeconds ?? 900) * 1000; // 15 minutes default
    const deadline = Math.min(
      Date.now() + response.expiresIn * 1000,
      Date.now() + maxWait
    );

    let interval = (response.interval ?? 5) * 1000;
    let consecutiveNetworkFailures = 0;
    const maxNetworkFailures = 5;

    while (Date.now() < deadline) {
      // Wait before polling
      await this.sleep(interval);

      try {
        const result = await this.pollToken(response.deviceCode);
        consecutiveNetworkFailures = 0; // Reset on success

        switch (result.status) {
          case "success":
            return result.tokens;

          case "pending":
            continue;

          case "slow_down":
            interval += result.newInterval * 1000;
            continue;

          case "expired":
            throw new AuthorizationError(
              "Device code expired",
              "DeviceCodeExpired"
            );

          case "access_denied":
            throw new AuthorizationError(
              "User denied authorization",
              "AccessDenied"
            );
        }
      } catch (error) {
        if (error instanceof AuthorizationError) {
          throw error;
        }

        // Network error - retry unless too many failures
        consecutiveNetworkFailures++;
        if (consecutiveNetworkFailures >= maxNetworkFailures) {
          throw error;
        }
        continue;
      }
    }

    throw new AuthorizationError("Device code expired", "DeviceCodeExpired");
  }

  private parseDeviceAuthorizationResponse(body: string): DeviceAuthorizationResponse {
    try {
      const data = JSON.parse(body);

      if (!data.device_code || !data.user_code || !data.verification_uri) {
        throw new ProviderError(
          "Invalid device authorization response",
          "InvalidRequest"
        );
      }

      return {
        deviceCode: data.device_code,
        userCode: data.user_code,
        verificationUri: data.verification_uri,
        verificationUriComplete: data.verification_uri_complete,
        expiresIn: data.expires_in ?? 1800, // Default 30 minutes
        interval: data.interval ?? 5,
      };
    } catch (error) {
      if (error instanceof OAuth2Error) {
        throw error;
      }
      throw new ProviderError(
        "Invalid device authorization response",
        "InvalidRequest"
      );
    }
  }

  private parseTokenResponse(body: string): TokenResponse {
    try {
      const data = JSON.parse(body);

      if (!data.access_token) {
        throw new ProviderError("Missing access_token in response", "InvalidRequest");
      }

      return {
        accessToken: data.access_token,
        tokenType: data.token_type ?? "Bearer",
        expiresIn: data.expires_in,
        refreshToken: data.refresh_token,
        scope: data.scope,
        idToken: data.id_token,
      };
    } catch (error) {
      if (error instanceof OAuth2Error) {
        throw error;
      }
      throw new ProviderError("Invalid token response", "InvalidRequest");
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

/**
 * Mock Device Authorization Flow for testing.
 */
export class MockDeviceAuthorizationFlow implements DeviceAuthorizationFlow {
  private requestDeviceCodeHistory: DeviceCodeParams[] = [];
  private pollHistory: string[] = [];
  private nextDeviceResponse?: DeviceAuthorizationResponse;
  private nextPollResult?: DeviceTokenResult;
  private pollCount: number = 0;
  private pollSuccessAfter: number = 3;

  /**
   * Set the next device authorization response.
   */
  setNextDeviceResponse(response: DeviceAuthorizationResponse): this {
    this.nextDeviceResponse = response;
    return this;
  }

  /**
   * Set the next poll result.
   */
  setNextPollResult(result: DeviceTokenResult): this {
    this.nextPollResult = result;
    return this;
  }

  /**
   * Set number of polls before success.
   */
  setPollSuccessAfter(count: number): this {
    this.pollSuccessAfter = count;
    return this;
  }

  /**
   * Get request device code history.
   */
  getRequestDeviceCodeHistory(): DeviceCodeParams[] {
    return [...this.requestDeviceCodeHistory];
  }

  /**
   * Get poll history.
   */
  getPollHistory(): string[] {
    return [...this.pollHistory];
  }

  async requestDeviceCode(params?: DeviceCodeParams): Promise<DeviceAuthorizationResponse> {
    this.requestDeviceCodeHistory.push(params ?? {});

    if (this.nextDeviceResponse) {
      const response = this.nextDeviceResponse;
      this.nextDeviceResponse = undefined;
      return response;
    }

    return {
      deviceCode: "mock-device-code",
      userCode: "MOCK-CODE",
      verificationUri: "https://mock.example.com/device",
      verificationUriComplete: "https://mock.example.com/device?user_code=MOCK-CODE",
      expiresIn: 1800,
      interval: 5,
    };
  }

  async pollToken(deviceCode: string): Promise<DeviceTokenResult> {
    this.pollHistory.push(deviceCode);
    this.pollCount++;

    if (this.nextPollResult) {
      const result = this.nextPollResult;
      this.nextPollResult = undefined;
      return result;
    }

    // Simulate pending until pollSuccessAfter polls
    if (this.pollCount < this.pollSuccessAfter) {
      return { status: "pending" };
    }

    return {
      status: "success",
      tokens: {
        accessToken: "mock-device-access-token",
        tokenType: "Bearer",
        expiresIn: 3600,
        refreshToken: "mock-device-refresh-token",
        scope: "openid profile",
      },
    };
  }

  async awaitAuthorization(
    response: DeviceAuthorizationResponse,
    _options?: { maxWaitSeconds?: number }
  ): Promise<TokenResponse> {
    // Simplified mock - just call pollToken once
    const result = await this.pollToken(response.deviceCode);
    if (result.status === "success") {
      return result.tokens;
    }
    throw new AuthorizationError("Mock authorization failed", "AccessDenied");
  }
}

/**
 * Create mock Device Authorization Flow for testing.
 */
export function createMockDeviceAuthorizationFlow(): MockDeviceAuthorizationFlow {
  return new MockDeviceAuthorizationFlow();
}

File: oauth2/rust/src/flows/device.rs
//! Device Authorization Flow
//!
//! RFC 8628 - OAuth 2.0 Device Authorization Grant.

use async_trait::async_trait;
use base64::Engine;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use crate::core::{HttpMethod, HttpRequest, HttpTransport};
use crate::error::{DeviceFlowError, OAuth2Error, ProtocolError};
use crate::types::{
    ClientAuthMethod, DeviceAuthorizationResponse, DeviceCodeParams, DeviceTokenResult,
    OAuth2Config, TokenResponse,
};
use crate::error::create_error_from_response;

/// Device Authorization Flow interface.
#[async_trait]
pub trait DeviceAuthorizationFlow: Send + Sync {
    /// Request device and user codes.
    async fn request_device_code(
        &self,
        params: DeviceCodeParams,
    ) -> Result<DeviceAuthorizationResponse, OAuth2Error>;

    /// Poll for token using device code.
    async fn poll_for_token(
        &self,
        device_code: &str,
    ) -> Result<DeviceTokenResult, OAuth2Error>;

    /// Poll for token with automatic retry until completion or expiration.
    async fn poll_until_complete(
        &self,
        device_code: &str,
        interval: Duration,
        expires_in: Duration,
    ) -> Result<TokenResponse, OAuth2Error>;
}

/// Device Authorization Flow implementation.
pub struct DeviceAuthorizationFlowImpl<T: HttpTransport> {
    config: OAuth2Config,
    transport: Arc<T>,
}

impl<T: HttpTransport> DeviceAuthorizationFlowImpl<T> {
    /// Create new Device Authorization Flow.
    pub fn new(config: OAuth2Config, transport: Arc<T>) -> Self {
        Self { config, transport }
    }

    fn build_device_code_request_body(&self, params: &DeviceCodeParams) -> String {
        let mut request_params = vec![];

        // Client ID (always required for device flow)
        request_params.push(("client_id", self.config.credentials.client_id.clone()));

        // Scopes
        let scopes = params
            .scopes
            .as_ref()
            .or(Some(&self.config.default_scopes));
        if let Some(s) = scopes {
            if !s.is_empty() {
                request_params.push(("scope", s.join(" ")));
            }
        }

        // Extra parameters
        for (key, value) in &params.extra_params {
            request_params.push((key.as_str(), value.clone()));
        }

        request_params
            .into_iter()
            .map(|(k, v)| format!("{}={}", urlencoding::encode(k), urlencoding::encode(&v)))
            .collect::<Vec<_>>()
            .join("&")
    }

    fn build_token_request_body(&self, device_code: &str) -> String {
        let mut params = vec![
            ("grant_type", "urn:ietf:params:oauth:grant-type:device_code".to_string()),
            ("device_code", device_code.to_string()),
            ("client_id", self.config.credentials.client_id.clone()),
        ];

        // Add client secret if using post method
        if self.config.credentials.auth_method == ClientAuthMethod::ClientSecretPost {
            if let Some(secret) = &self.config.credentials.client_secret {
                use secrecy::ExposeSecret;
                params.push(("client_secret", secret.expose_secret().to_string()));
            }
        }

        params
            .into_iter()
            .map(|(k, v)| format!("{}={}", urlencoding::encode(k), urlencoding::encode(&v)))
            .collect::<Vec<_>>()
            .join("&")
    }

    fn build_request_headers(&self) -> HashMap<String, String> {
        let mut headers = HashMap::new();
        headers.insert(
            "content-type".to_string(),
            "application/x-www-form-urlencoded".to_string(),
        );
        headers.insert("accept".to_string(), "application/json".to_string());

        // Add Basic auth header if using that method
        if self.config.credentials.auth_method == ClientAuthMethod::ClientSecretBasic {
            if let Some(secret) = &self.config.credentials.client_secret {
                use secrecy::ExposeSecret;
                let credentials = format!(
                    "{}:{}",
                    self.config.credentials.client_id,
                    secret.expose_secret()
                );
                let encoded = base64::engine::general_purpose::STANDARD.encode(credentials);
                headers.insert("authorization".to_string(), format!("Basic {}", encoded));
            }
        }

        headers
    }

    fn parse_poll_error(&self, body: &str) -> DeviceTokenResult {
        // Try to parse as OAuth error response
        #[derive(serde::Deserialize)]
        struct ErrorResponse {
            error: String,
            #[serde(default)]
            error_description: Option<String>,
        }

        if let Ok(error_resp) = serde_json::from_str::<ErrorResponse>(body) {
            match error_resp.error.as_str() {
                "authorization_pending" => DeviceTokenResult::Pending,
                "slow_down" => DeviceTokenResult::SlowDown,
                "access_denied" => DeviceTokenResult::AccessDenied {
                    error_description: error_resp.error_description,
                },
                "expired_token" => DeviceTokenResult::Expired,
                _ => DeviceTokenResult::Error(OAuth2Error::DeviceFlow(
                    DeviceFlowError::DeviceCodeError {
                        error: error_resp.error,
                        error_description: error_resp.error_description,
                    },
                )),
            }
        } else {
            DeviceTokenResult::Error(OAuth2Error::Protocol(ProtocolError::InvalidJson {
                message: "Failed to parse error response".to_string(),
            }))
        }
    }
}

#[async_trait]
impl<T: HttpTransport> DeviceAuthorizationFlow for DeviceAuthorizationFlowImpl<T> {
    async fn request_device_code(
        &self,
        params: DeviceCodeParams,
    ) -> Result<DeviceAuthorizationResponse, OAuth2Error> {
        // Get device authorization endpoint
        let device_endpoint = self
            .config
            .provider
            .device_authorization_endpoint
            .as_ref()
            .ok_or_else(|| {
                OAuth2Error::DeviceFlow(DeviceFlowError::DeviceFlowNotSupported)
            })?;

        let body = self.build_device_code_request_body(&params);
        let headers = self.build_request_headers();

        let http_request = HttpRequest {
            method: HttpMethod::Post,
            url: device_endpoint.clone(),
            headers,
            body: Some(body),
            timeout: Some(self.config.timeout),
        };

        let response = self.transport.send(http_request).await?;

        if response.status != 200 {
            return Err(create_error_from_response(response.status, &response.body));
        }

        let device_response: DeviceAuthorizationResponse =
            serde_json::from_str(&response.body).map_err(|e| {
                OAuth2Error::Protocol(ProtocolError::InvalidJson {
                    message: e.to_string(),
                })
            })?;

        Ok(device_response)
    }

    async fn poll_for_token(&self, device_code: &str) -> Result<DeviceTokenResult, OAuth2Error> {
        let body = self.build_token_request_body(device_code);
        let headers = self.build_request_headers();

        let http_request = HttpRequest {
            method: HttpMethod::Post,
            url: self.config.provider.token_endpoint.clone(),
            headers,
            body: Some(body),
            timeout: Some(self.config.timeout),
        };

        let response = self.transport.send(http_request).await?;

        if response.status == 200 {
            let token_response: TokenResponse =
                serde_json::from_str(&response.body).map_err(|e| {
                    OAuth2Error::Protocol(ProtocolError::InvalidJson {
                        message: e.to_string(),
                    })
                })?;
            return Ok(DeviceTokenResult::Success(token_response));
        }

        // Handle polling errors (400 status is expected during polling)
        if response.status == 400 {
            return Ok(self.parse_poll_error(&response.body));
        }

        Err(create_error_from_response(response.status, &response.body))
    }

    async fn poll_until_complete(
        &self,
        device_code: &str,
        interval: Duration,
        expires_in: Duration,
    ) -> Result<TokenResponse, OAuth2Error> {
        let start = std::time::Instant::now();
        let mut current_interval = interval;

        loop {
            // Check for expiration
            if start.elapsed() >= expires_in {
                return Err(OAuth2Error::DeviceFlow(DeviceFlowError::DeviceCodeExpired));
            }

            // Wait before polling
            tokio::time::sleep(current_interval).await;

            match self.poll_for_token(device_code).await? {
                DeviceTokenResult::Success(token) => return Ok(token),
                DeviceTokenResult::Pending => {
                    // Continue polling at current interval
                    continue;
                }
                DeviceTokenResult::SlowDown => {
                    // Increase interval by 5 seconds as per RFC 8628
                    current_interval += Duration::from_secs(5);
                    continue;
                }
                DeviceTokenResult::AccessDenied { error_description } => {
                    return Err(OAuth2Error::DeviceFlow(DeviceFlowError::DeviceCodeError {
                        error: "access_denied".to_string(),
                        error_description,
                    }));
                }
                DeviceTokenResult::Expired => {
                    return Err(OAuth2Error::DeviceFlow(DeviceFlowError::DeviceCodeExpired));
                }
                DeviceTokenResult::Error(e) => return Err(e),
            }
        }
    }
}

/// Mock Device Authorization Flow for testing.
#[derive(Default)]
pub struct MockDeviceAuthorizationFlow {
    device_code_history: std::sync::Mutex<Vec<DeviceCodeParams>>,
    poll_history: std::sync::Mutex<Vec<String>>,
    next_device_response: std::sync::Mutex<Option<DeviceAuthorizationResponse>>,
    next_poll_result: std::sync::Mutex<Option<DeviceTokenResult>>,
    next_error: std::sync::Mutex<Option<OAuth2Error>>,
    poll_count: std::sync::Mutex<u32>,
    polls_until_success: std::sync::Mutex<Option<u32>>,
}

impl MockDeviceAuthorizationFlow {
    /// Create new mock flow.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set next device code response.
    pub fn set_next_device_response(&self, response: DeviceAuthorizationResponse) -> &Self {
        *self.next_device_response.lock().unwrap() = Some(response);
        self
    }

    /// Set next poll result.
    pub fn set_next_poll_result(&self, result: DeviceTokenResult) -> &Self {
        *self.next_poll_result.lock().unwrap() = Some(result);
        self
    }

    /// Set next error.
    pub fn set_next_error(&self, error: OAuth2Error) -> &Self {
        *self.next_error.lock().unwrap() = Some(error);
        self
    }

    /// Set number of polls until success.
    pub fn set_polls_until_success(&self, count: u32) -> &Self {
        *self.polls_until_success.lock().unwrap() = Some(count);
        self
    }

    /// Get device code request history.
    pub fn get_device_code_history(&self) -> Vec<DeviceCodeParams> {
        self.device_code_history.lock().unwrap().clone()
    }

    /// Get poll history.
    pub fn get_poll_history(&self) -> Vec<String> {
        self.poll_history.lock().unwrap().clone()
    }

    /// Get poll count.
    pub fn get_poll_count(&self) -> u32 {
        *self.poll_count.lock().unwrap()
    }
}

#[async_trait]
impl DeviceAuthorizationFlow for MockDeviceAuthorizationFlow {
    async fn request_device_code(
        &self,
        params: DeviceCodeParams,
    ) -> Result<DeviceAuthorizationResponse, OAuth2Error> {
        self.device_code_history.lock().unwrap().push(params);

        if let Some(error) = self.next_error.lock().unwrap().take() {
            return Err(error);
        }

        if let Some(response) = self.next_device_response.lock().unwrap().take() {
            return Ok(response);
        }

        Ok(DeviceAuthorizationResponse {
            device_code: "mock-device-code".to_string(),
            user_code: "ABCD-EFGH".to_string(),
            verification_uri: "https://example.com/device".to_string(),
            verification_uri_complete: Some(
                "https://example.com/device?user_code=ABCD-EFGH".to_string(),
            ),
            expires_in: 1800,
            interval: Some(5),
        })
    }

    async fn poll_for_token(&self, device_code: &str) -> Result<DeviceTokenResult, OAuth2Error> {
        self.poll_history
            .lock()
            .unwrap()
            .push(device_code.to_string());

        let mut count = self.poll_count.lock().unwrap();
        *count += 1;

        if let Some(error) = self.next_error.lock().unwrap().take() {
            return Err(error);
        }

        if let Some(result) = self.next_poll_result.lock().unwrap().take() {
            return Ok(result);
        }

        // Check if we should return success after N polls
        if let Some(polls_needed) = *self.polls_until_success.lock().unwrap() {
            if *count >= polls_needed {
                return Ok(DeviceTokenResult::Success(TokenResponse {
                    access_token: "mock-device-token".to_string(),
                    token_type: "Bearer".to_string(),
                    expires_in: Some(3600),
                    refresh_token: Some("mock-refresh-token".to_string()),
                    scope: Some("openid profile".to_string()),
                    id_token: None,
                    extra: HashMap::new(),
                }));
            }
        }

        Ok(DeviceTokenResult::Pending)
    }

    async fn poll_until_complete(
        &self,
        device_code: &str,
        interval: Duration,
        expires_in: Duration,
    ) -> Result<TokenResponse, OAuth2Error> {
        let start = std::time::Instant::now();
        let mut current_interval = interval;

        loop {
            if start.elapsed() >= expires_in {
                return Err(OAuth2Error::DeviceFlow(DeviceFlowError::DeviceCodeExpired));
            }

            tokio::time::sleep(current_interval).await;

            match self.poll_for_token(device_code).await? {
                DeviceTokenResult::Success(token) => return Ok(token),
                DeviceTokenResult::Pending => continue,
                DeviceTokenResult::SlowDown => {
                    current_interval += Duration::from_secs(5);
                    continue;
                }
                DeviceTokenResult::AccessDenied { error_description } => {
                    return Err(OAuth2Error::DeviceFlow(DeviceFlowError::DeviceCodeError {
                        error: "access_denied".to_string(),
                        error_description,
                    }));
                }
                DeviceTokenResult::Expired => {
                    return Err(OAuth2Error::DeviceFlow(DeviceFlowError::DeviceCodeExpired));
                }
                DeviceTokenResult::Error(e) => return Err(e),
            }
        }
    }
}

/// Create mock Device Authorization Flow for testing.
pub fn create_mock_device_authorization_flow() -> MockDeviceAuthorizationFlow {
    MockDeviceAuthorizationFlow::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_mock_device_code_request() {
        let flow = MockDeviceAuthorizationFlow::new();

        let params = DeviceCodeParams {
            scopes: Some(vec!["openid".to_string()]),
            ..Default::default()
        };

        let response = flow.request_device_code(params).await.unwrap();
        assert_eq!(response.device_code, "mock-device-code");
        assert_eq!(response.user_code, "ABCD-EFGH");
        assert!(response.verification_uri_complete.is_some());

        let history = flow.get_device_code_history();
        assert_eq!(history.len(), 1);
    }

    #[tokio::test]
    async fn test_mock_poll_pending() {
        let flow = MockDeviceAuthorizationFlow::new();

        let result = flow.poll_for_token("test-device-code").await.unwrap();
        assert!(matches!(result, DeviceTokenResult::Pending));

        let history = flow.get_poll_history();
        assert_eq!(history.len(), 1);
        assert_eq!(history[0], "test-device-code");
    }

    #[tokio::test]
    async fn test_mock_poll_until_success() {
        let flow = MockDeviceAuthorizationFlow::new();
        flow.set_polls_until_success(2);

        // First poll should be pending
        let result = flow.poll_for_token("test-code").await.unwrap();
        assert!(matches!(result, DeviceTokenResult::Pending));

        // Second poll should succeed
        let result = flow.poll_for_token("test-code").await.unwrap();
        match result {
            DeviceTokenResult::Success(token) => {
                assert_eq!(token.access_token, "mock-device-token");
            }
            _ => panic!("Expected success"),
        }

        assert_eq!(flow.get_poll_count(), 2);
    }

    #[tokio::test]
    async fn test_mock_custom_device_response() {
        let flow = MockDeviceAuthorizationFlow::new();
        flow.set_next_device_response(DeviceAuthorizationResponse {
            device_code: "custom-code".to_string(),
            user_code: "CUSTOM".to_string(),
            verification_uri: "https://custom.example.com/device".to_string(),
            verification_uri_complete: None,
            expires_in: 900,
            interval: Some(10),
        });

        let params = DeviceCodeParams::default();
        let response = flow.request_device_code(params).await.unwrap();
        assert_eq!(response.device_code, "custom-code");
        assert_eq!(response.expires_in, 900);
    }
}
